### ISSUE TITLE
Decreased Upsert Performance

### ISSUE THREAD

**Issue by markbigler** [bwaldvogel/mongo-java-server#84]
We store the communication (requests, responses and durations per minute) between all our applications in a statistics collection. To improve the update performance, "empty" documents are pre-generated (example update query see below) for the next days.

With the update from version 1.17.1 to 1.18, the performance of the following update query has increased from ~4 seconds to ~11 seconds:

```
{
  "$inc": {
    "version": 1,
    "data.0.0.requests": 0,
    "data.0.0.responses": 0,
    "data.0.0.duration": 0,
    "data.0.1.requests": 0,
    "data.0.1.responses": 0,
    "data.0.1.duration": 0,

    ... foreach hour.minute ...

    "data.23.59.requests": 0,
    "data.23.59.responses": 0,
    "data.23.59.duration": 0
  },
  "$setOnInsert": {
    "application": "application-name",
    "tenant": "tenant-name",
    "name": "name",
    "timestamp": {
      "$date": 1462658400000
    }
  },
  "$set": {
    "preAllocated": true
  }
}
```

**Comment by bwaldvogel** [comment]
Thanks for reporting this performance regression. It’s fixed in version `1.18.1`.

**Comment by markbigler** [comment]
Wow, that went incredibly fast! 👏

I did a retest with 1.18.1 and the performance (~1 second) is even better than with 1.17.1. Thanks for the fix.
